### DISCUSSION TITLE
How to write a `Wait` or `Ensure` that "or"s several expectations together?

### DISCUSSION THREAD

**Discussion by perrygoy** [serenity-js/serenity-js#2309] [How to... ?] [answered]
I have a search bar that will either display search results or a "results not found" message. I'm trying to write a task which will commit a search and wait for one of those two things to happen.

How can i write a `Wait` or `Ensure` that will allow me to say "either at least one search result appears or the 'results not found' message appears"? I saw [the documentation](https://serenity-js.org/api/assertions/class/Ensure/) where one "actual" is compared against several expectations with `and`, but that's not quite what i want. Something like:
```javascript
theActor.attemptsTo(
    Ensure.thatAtLeastOneOf(
        (Search.searchResult(), isPresent()),
        (Search.noResultsMessage(), isPresent()),
    )
)
```

**Comment by jan-molak** [comment, marked as answer]
It seems like you can either get a `results not found` message, or a search result, but never both simultaneously. If so, I'd suggest using a [conditional statement](https://serenity-js.org/handbook/design/control-flow/#conditional-statements) to [`Check`](https://serenity-js.org/api/core/class/Check/) if the results are present, and if they're not - look for the `results not found` message.

Something along those lines:

```ts
import { Check } from '@serenty-js/core'
import { Ensure. isPresent } from '@serenity-js/assertions'
import { Text } from '@serenity-js/web'

const VerifySearchResults = () => 
  Task.where(`#actor verifies search results`,
    Check.whether(Search.noResultsMessage(), isPresent())
      .andIfSo(
        Ensure.that(Text.of(Search.noResultsMessage()), equals('results not found'))
      )
      .otherwise(
        Ensure.that(Search.searchResult(), isPresent()),
      )
  )
```

Hope this helps!

**Reply by perrygoy** [reply]
This will work, but it will be a little flakey, because i'll need to wait for the "search in progress" loading animation to appear and disappear. In my experience with these tests, sometimes that search loader will appear and disappear so quickly that the test doesn't have time to check for it. 

It would be really nice to be able to tap into the `Ensure.eventually` mechanism to ensure at least one of these conditions is met _at some point._ But i do think this approach will mostly work. 

Thank you for your help!
